datasource/src/main/scala/quasar/blobstore/azure/AzureBlobstore.scala

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.blobstore.azure
+
+import slamdata.Predef._
+import quasar.api.resource.{ResourceName, ResourcePath, ResourcePathType}
+import quasar.blobstore.Blobstore
+import quasar.contrib.std.errorNotImplemented
+
+import java.lang.Integer
+import scala.collection.JavaConverters._
+
+import cats.effect._
+import cats.effect.concurrent.Deferred
+import cats.implicits._
+import com.microsoft.azure.storage.blob._
+import com.microsoft.azure.storage.blob.models._
+import com.microsoft.rest.v2.Context
+import fs2.Stream
+import io.reactivex._
+import io.reactivex.disposables.Disposable
+import io.reactivex.observers._
+
+object AzureBlobstore {
+
+  final class AsyncObserver[A](cb: Either[Throwable, A] => Unit)
+    extends DisposableSingleObserver[A] {
+
+    override def onStart(): Unit = ()
+
+    override def onSuccess(a: A) =
+      cb(Right(a))
+
+    override def onError(t: Throwable) =
+      cb(Left(t))
+  }
+}
+
+class AzureBlobstore[F[_]: Concurrent](
+  containerURL: ContainerURL) extends Blobstore[F] {
+
+  import AzureBlobstore._
+
+  private val F = Async[F]
+
+  def get(path: ResourcePath): Stream[F, Byte] = errorNotImplemented
+
+  def isResource(path: ResourcePath): F[Boolean] = errorNotImplemented
+
+  def list(path: ResourcePath): Stream[F, (ResourceName, ResourcePathType)] = {
+    val resp: F[ContainerListBlobHierarchySegmentResponse] = for {
+      single <- listBlobs(None, pathToOptions(path))
+      d <- Deferred[F, Disposable]
+      r <- mkAsync(d, single.subscribe)
+      disposable <- d.get
+      _ <- F.delay(disposable.dispose)
+    } yield r
+
+    Stream.eval(resp).flatMap { r =>
+      val segm = r.body.segment
+      val l =
+        if (segm == null) List.empty
```

Comment:
this is the not found case. Maybe we should raise a `ResourceError.PathNotFound` here?

Reply:
note that any way this would need to be augmented:
https://github.com/slamdata/quasar/blob/master/connector/src/test/scala/quasar/connector/DatasourceSpec.scala#L46-L50

Reply:
@wemrysi I think I'd prefer raising `ResourceError.PathNotFound` here because that more precisely represents the situation, but I'm not sure if this is according to the design. Basically this means that the test referred to above will have both an out-of-stream and an in-stream representation of not found.

Reply:
Why not return `None`? That is the contract for `Datasource`.

Reply:
If you don’t `Stream.eval` on L73 then you wouldn’t need to `sequence`.

Reply:
Ah yes!